src/storage/gc_worker.rs

```diff
@@ -0,0 +1,306 @@
+    }
+}
+
+/// `GCRunner` is used to perform GC on the engine
+struct GCRunner<E: Engine> {
+    engine: E,
+    ratio_threshold: f64,
+
+    stats: StatisticsSummary,
+}
+
+impl<E: Engine> GCRunner<E> {
+    pub fn new(engine: E, ratio_threshold: f64) -> GCRunner<E> {
+        GCRunner {
+            engine,
+            ratio_threshold,
+
+            stats: StatisticsSummary::default(),
+        }
+    }
+
+    fn get_snapshot(&self, ctx: &mut Context) -> Result<E::Snap> {
+        let timeout = Duration::from_secs(GC_SNAPSHOT_TIMEOUT_SECS);
+        match wait_op!(|cb| self.engine.async_snapshot(ctx, cb), timeout) {
+            Some((cb_ctx, Ok(snapshot))) => {
+                if let Some(term) = cb_ctx.term {
+                    ctx.set_term(term);
+                }
+                Ok(snapshot)
+            }
+            Some((_, Err(e))) => Err(e),
+            None => Err(EngineError::Timeout(timeout)),
+        }.map_err(Error::from)
+    }
+
+    /// Scan keys in the region. Returns scanned keys if any, and a key indicating scan progress
+    fn scan_keys(
+        &mut self,
+        ctx: &mut Context,
+        safe_point: u64,
+        from: Option<Key>,
+        limit: usize,
+    ) -> Result<(Vec<Key>, Option<Key>)> {
+        let snapshot = self.get_snapshot(ctx)?;
+        let mut reader = MvccReader::new(
+            snapshot,
+            Some(ScanMode::Forward),
+            !ctx.get_not_fill_cache(),
+            None,
+            None,
+            ctx.get_isolation_level(),
+        );
+
+        // If from.is_some(), it must not be the first scan of the region.
+        // So we must continue doing GC.
+        let skip_gc = from.is_none() && !reader.need_gc(safe_point, self.ratio_threshold);
+        let res = if skip_gc {
+            KV_GC_SKIPPED_COUNTER.inc();
+            Ok((vec![], None))
+        } else {
+            reader
+                .scan_keys(from.clone(), limit)
+                .map_err(Error::from)
+                .and_then(|(keys, next)| {
+                    if keys.is_empty() {
+                        assert!(next.is_none());
+                        if from.is_none() {
+                            KV_GC_EMPTY_RANGE_COUNTER.inc();
+                        }
+                    }
+                    Ok((keys, next))
+                })
+        };
+        self.stats.add_statistics(reader.get_statistics());
+        res
+    }
+
+    /// Clean up outdated data.
+    fn gc_keys(
+        &mut self,
+        ctx: &mut Context,
+        safe_point: u64,
+        keys: Vec<Key>,
+        mut next_scan_key: Option<Key>,
+    ) -> Result<Option<Key>> {
+        let snapshot = self.get_snapshot(ctx)?;
+        let mut txn = MvccTxn::new(
+            snapshot,
+            0,
+            Some(ScanMode::Forward),
+            ctx.get_isolation_level(),
+            !ctx.get_not_fill_cache(),
+        );
+        for k in keys {
+            // TODO: Duplicated code in scheduler.rs
```

Comment:
Are there some reasons to keep the code of the scheduler's gc?

Reply:
No. They will be removed from scheduler later, or if you think it's necessary, I can remove them in this PR

Reply:
Yes I think it can be done in this PR.

Reply:
since @zhangjinpeng1987 has already given approvals, I think you can also do it in another PR.